embassy-rp/src/i2c_slave.rs

```diff
@@ -161,53 +236,139 @@ impl<'d, T: Instance> I2cSlave<'d, T> {
         }
     }
 
-    /// Wait asynchronously for commands from an I2C master.
-    /// `buffer` is provided in case master does a 'write' and is unused for 'read'.
-    pub async fn listen(&mut self, buffer: &mut [u8]) -> Result<Command, Error> {
+    #[inline(always)]
+    fn rx_not_empty(&self) -> bool {
         let p = T::regs();
+        self.pending_byte.is_some() || p.ic_status().read().rfne()
+    }
 
-        p.ic_clr_intr().read();
-        // set rx fifo watermark to 1 byte
-        p.ic_rx_tl().write(|w| w.set_rx_tl(0));
+    #[inline(always)]
+    async fn next_event(&mut self) -> I2cSlaveEvent {
```

Comment:
I don't think this state makes the logic clearer.

- There's events that are unused (start, restart...). 
- At start of `listen`, we know the bus state is always idle. Therefore there should be no need to save the state in `I2cSlave`. Saving it means at start of `listen` the state could be anything now. This makes reasoning about the core harder. 
- There's now two "layered" state machines, `next_event` and `listen`.

Could you do the required changes for the sisues you found without this refactor?  Thank you!